File: UVa Online Judge/11013.py
# =============================================================================
#  Author:          Teerapat Jenrungrot - https://github.com/mjenrungrot/
#  FileName:        11013.py
#  Description:     UVa Online Judge - 11013
# =============================================================================
import sys

def get_rank(card):
    if card[0] == 'A': return 0
    if card[0] == 'T': return 9
    if card[0] == 'J': return 10
    if card[0] == 'Q': return 11
    if card[0] == 'K': return 12
    return ord(card[0]) - ord('1')

memory = {}

def evaluate(cards):
    ranks = list(map(get_rank, cards))
    sorted_ranks = tuple(sorted(ranks))

    if sorted_ranks in memory:
        return memory[sorted_ranks]

    bins = [0 for i in range(13)]
    for rank in ranks:
        bins[rank] += 1

    start = 0
    while bins[start] > 0:
        start = (start + 1) % 13
    while bins[start] == 0:
        start = (start + 1) % 13

    consec_bins = []
    n_cards = 0
    counter = 0
    curr = start
    while n_cards < 5:
        if bins[curr] > 0:
            bins[curr] -= 1
            counter += 1
        else:
            if counter > 0:
                consec_bins.append(counter)
                n_cards += counter
            counter = 0
        curr = (curr + 1) % 13
    consec_bins.sort(reverse=True)
    best_score = 0
    if consec_bins[0] == 5: best_score = 100
    elif consec_bins[0] == 4: best_score = 10
    elif len(consec_bins) >= 2 and consec_bins[0] == 3 and consec_bins[1] == 2: best_score = 5
    elif consec_bins[0] == 3: best_score = 3
    elif len(consec_bins) >= 2 and consec_bins[0] == 2 and consec_bins[1] == 2: best_score = 1
    memory[sorted_ranks] = best_score
    return best_score

if __name__ == '__main__':
    all_cards = ["{}{}".format(rank, suit) for rank in "A23456789TJQK" for suit in "SHDC"]
    while True:
        line = sys.stdin.readline()
        if line.strip() == "#":
            break
        cards = line.strip().split()

        remaining_cards = list(filter(lambda x: x not in cards, all_cards[:]))
        best_expectation = -1e9
        best_choice = None

        curr_money = evaluate(cards)

        for idx in range(len(cards)):
            expectation = -47
            for rem_card in remaining_cards:
                new_decks = cards[:]
                new_decks[idx] = rem_card
                score = evaluate(new_decks)
                expectation += score
            expectation /= 47.0

            if expectation > curr_money and expectation > best_expectation:
                best_expectation = expectation
                best_choice = cards[idx]

        if best_choice:
            print("Exchange {}".format(best_choice))
        else:
            print("Stay")

File: UVa Online Judge/11013.cc
#include <cstdio>
#include <vector>
#include <string>
#include <iostream>
#include <algorithm>
using namespace std;


vector<string> get_remaining_cards(vector<string> cards){
    const string rank = "A23456789TJQK";
    const string suit = "SHDC";
    vector <string> x;
    for(auto it1=rank.cbegin(); it1!=rank.cend(); it1++){
        for(auto it2=suit.cbegin(); it2!=suit.cend(); it2++){
            string candidate = "";
            candidate = candidate + *it1 + *it2;
            bool found = false;
            for(auto card: cards){
                if(candidate == card){
                    found = true;
                    break;
                }
            }
            if(not found) x.push_back(candidate);
        }
    }
    return x;
}

int get_rank(string card){
    if(card[0] == 'A') return 0;
    else if(card[0] >= '2' and card[0] <= '9') return card[0] - '1';
    else if(card[0] == 'T') return 9;
    else if(card[0] == 'J') return 10;
    else if(card[0] == 'Q') return 11;
    else return 12;
}

int next_rank(int rank){
    return (rank + 1) % 13;
}

int evaluate(vector<string> cc){
    vector <int> ranks;
    for(int i=0; i<5; i++){
        ranks.push_back(get_rank(cc[i]));
    }
    if(next_rank(ranks[0]) == ranks[1]){
        if(next_rank(ranks[1]) == ranks[2]){
            if(next_rank(ranks[2]) == ranks[3]){
                if(next_rank(ranks[3]) == ranks[4]) return 100;
                return 10;
            }
            if(next_rank(ranks[3]) == ranks[4]) return 5;
            return 3;
        }
        if(next_rank(ranks[2]) == ranks[3]) return 1;
        return 0;
    }
    return 0;
}

int evaluate_wrapper(vector<string> cc){
    int ans = 0;
    sort(cc.begin(), cc.end());
    do{
        ans = max(ans, evaluate(cc));
    }while(next_permutation(cc.begin(), cc.end()));
    return ans;
}

int main(){
    string card;
    vector <string> cards;
    while(cin >> card){
        if(card == "#") break;
        cards.clear();
        cards.push_back(card);
        for(int i=1; i<5; i++){
            cin >> card;
            cards.push_back(card);
        }
        int curr_score = evaluate_wrapper(cards);
        vector<string> rem_cards = get_remaining_cards(cards);
        vector<string> tmp_cards = cards;

        double expectation, best_expectation = -1e9;
        int best_card, score;
        for(int i=0; i<5; i++){
            expectation = -curr_score - 1.0;
            for(auto x: rem_cards){
                tmp_cards[i] = x;
                score = evaluate_wrapper(tmp_cards);
                expectation += 1.0 / 47.0 * score;
            }
            tmp_cards[i] = cards[i];

            if(expectation > best_expectation){
                best_expectation = expectation;
                best_card = i;
            }
        }

        if(best_expectation < 0.0) printf("Stay\n");
        else printf("Exchange %s\n", cards[best_card].c_str());
    }
    return 0;
}

// TH JH QH KH AH
// 2H TC 5D 3S 6C